datadotworld/util.py

```diff
@@ -132,28 +132,39 @@ class memoized(object):
     If called later with the same arguments, the cached value is returned
     (not reevaluated).
     """
-
     def __init__(self, key_mapper=None):
         self.key_mapper = key_mapper
         self.cache = {}
 
     def __call__(self, func):
-        def wrapper(*args):
+        def wrapper(*args, **kwargs):
             """
 
             :param *args:
             """
             key = self.key_mapper(*args) or args
+            # get the instance of self so that when a class
+            # is re-constructed, we can clear the cache object.
+            obj = args[0]
+            obj_id = id(obj)
+
             if not isinstance(key, collections.Hashable):
                 # uncacheable. a list, for instance.
                 # better to not cache than blow up.
                 return func(*args)
-            if key in self.cache:
-                return self.cache[key]
-            else:
-                value = func(*args)
-                self.cache[key] = value
-                return value
+
+            # clear cache if object id does not exist in cache
+            if obj_id not in self.cache:
+                self.cache.clear()
```

Comment:
This is unnecessarily limiting, imho. This approach limits the benefits of caching if any two objects that use memoize are used concurrently even when they are instances of different classes. Instead, why not bind the cache to the object itself?

In other words, instead of:
```python
self.cache[obj_id] = self.cache.get(obj_id, {})
```
You can:
```python
try:
  obj.__memoized__ = getattr(obj, '__memoized__', {})
except AttributeError:
  raise TypeError('Unable to set property __memoized__ on object of type {}.'.format(type(obj)))
```

Reply:
This is also an interesting approach and I think binding the cache to the object will still retain the old cache bound to the object, so when load_dataset is called again, it's still going to refer to the old cache. If we have to bind to the object, we might need to keep track of objects created after every instantiation with some form of meta-programming and then clear the last cache on every load dataset.

Since every cached object is now unique, it's definitely unnecessary to clear cache, so removing lines, 157 and 158 will still do the trick.